hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java

```diff
@@ -109,13 +110,32 @@ CompositeType getComponentIdType() {
 	}
 
 	private Object readResolve() {
-		BytecodeProvider bytecodeProvider = Environment.getBytecodeProvider();
-		if ( !( bytecodeProvider instanceof BytecodeProviderImpl ) ) {
-			throw new IllegalStateException( "The bytecode provider is not ByteBuddy, unable to deserialize a ByteBuddy proxy." );
-		}
-
-		HibernateProxy proxy = ( (BytecodeProviderImpl) bytecodeProvider ).getByteBuddyProxyHelper().deserializeProxy( this );
+		final SessionFactoryImplementor sessionFactory = retrieveMatchingSessionFactory( this.sessionFactoryUuid );
+		BytecodeProviderImpl byteBuddyBytecodeProvider = retrieveByteBuddyBytecodeProvider( sessionFactory );
+		HibernateProxy proxy = byteBuddyBytecodeProvider.getByteBuddyProxyHelper().deserializeProxy( this );
 		afterDeserialization( (ByteBuddyInterceptor) proxy.getHibernateLazyInitializer() );
 		return proxy;
 	}
+
+	private static SessionFactoryImplementor retrieveMatchingSessionFactory(final String sessionFactoryUuid) {
+		Objects.requireNonNull( sessionFactoryUuid );
```

Comment:
We won't be able to deserialize entities serialized before this patch, since the uuid will be missing.

Not sure what the backwards compatibility expectations of serialization are, though.

Reply:
The UUID was always supposed to be there in practice - I only noticed this wasn't happening consistently because we have some proxy serialization tests which were performing this while in the odd configuration of not having set the `sessionFactoryUuid` first.
However in a "real" scenario this would have failed as the proxy can't be initialized.